File: spring_server/src/main/java/com/semothon/spring_server/room/dto/GetRoomResponseDto.java
package com.semothon.spring_server.room.dto;

import com.semothon.spring_server.room.entity.Room;
import lombok.*;

import java.util.List;
import java.util.stream.Collectors;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class GetRoomResponseDto {
    private RoomInfoDto roomInfo;
    private List<RoomUserInfoDto> members;
    private HostUserInfoDto host;

    public static GetRoomResponseDto from(Room room){
        return GetRoomResponseDto.builder()
                .roomInfo(RoomInfoDto.from(room))
                .members(room.getRoomUsers().stream()
                        .map(RoomUserInfoDto::from)
                        .collect(Collectors.toList()))
                .host(HostUserInfoDto.from(room.getHost()))
                .build();
    }
}

File: flutter_app/lib/dto/get_room_response_dto.dart
import 'room_info_dto.dart';
import 'room_user_info_dto.dart';
import 'host_user_info_dto.dart';

class GetRoomResponseDto {
  final RoomInfoDto roomInfo;
  final List<RoomUserInfoDto> members;
  final HostUserInfoDto host;

  GetRoomResponseDto({
    required this.roomInfo,
    required this.members,
    required this.host,
  });

  factory GetRoomResponseDto.fromJson(Map<String, dynamic> json) {
    return GetRoomResponseDto(
      roomInfo: RoomInfoDto.fromJson(json['roomInfo']),
      members: List<Map<String, dynamic>>.from(json['members'])
          .map((e) => RoomUserInfoDto.fromJson(e))
          .toList(),
      host: HostUserInfoDto.fromJson(json['host']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'roomInfo': roomInfo.toJson(),
      'members': members.map((e) => e.toJson()).toList(),
      'host': host.toJson(),
    };
  }
}
